binary_transparency/firmware/internal/verify/bundle.go

```diff
@@ -57,6 +57,39 @@ func BundleForUpdate(bundleRaw, fwHash []byte, dc api.LogCheckpoint, cpFunc Cons
 	return nil
 }
 
+// BundleValidateWitness verifies the received bundle against witness checkpoint
+func BundleValidateWitness(bundleRaw []byte, wc api.LogCheckpoint, cpFunc ConsistencyProofFunc) error {
+	var pb api.ProofBundle
+	lv := NewLogVerifier()
+	if err := json.Unmarshal(bundleRaw, &pb); err != nil {
+		return fmt.Errorf("failed to parse proof bundle: %w", err)
+	}
+	fmt.Printf("Witness TreeSize=%d, Inclusion Index=%d \n", wc.TreeSize, pb.InclusionProof.LeafIndex)
+
+	if wc.TreeSize < pb.InclusionProof.LeafIndex {
+		return fmt.Errorf("witness verification failed wcp treesize(%d)<device cp index(%d)", wc.TreeSize, pb.InclusionProof.LeafIndex)
+	}
+
+	fromts := wc.TreeSize
+	tots := pb.Checkpoint.TreeSize
+	root1 := wc.RootHash
+	root2 := pb.Checkpoint.RootHash
+	if wc.TreeSize > pb.Checkpoint.TreeSize {
+		fromts = pb.Checkpoint.TreeSize
+		tots = wc.TreeSize
+		root1 = pb.Checkpoint.RootHash
+		root2 = wc.RootHash
+	}
```

Comment:
I think it's worth adding a couple of tests for the various orderings of:
 * bundle inclusion index
 * bundle cp
 * witness cp

I've checked the code here and it looks fine, but this block in particular would be easy to break during future changes.

Reply:
As discussed during F2F conversation, agreed to slightly simplify the coding style to better readability and understanding!
Uploaded changes with new implementation!